src/com/providertrust/im981/Main.java

```diff
@@ -171,71 +191,144 @@ private static void reorderRecord(ArrayList<Tuple> tuples, PreparedStatement sel
                 if (auditResult.wasNull())
                     orderId = null;
                 int rowNumber = auditResult.getInt(5);
-                Tuple tuple = new Tuple(addressLinesId, line, orderId, rowNumber);
-                switch (revType)
-                {
-                    case 0:
-                        tuples.add(tuple);
-                        int order = tuples.size() - 1; // 0 based
-                        if (orderId == null || orderId != order)
-                        {
-                            tuple.setOrderId(order);
-                            auditResult.updateInt(ORDER_COLUMN, order);
-                            auditResult.updateRow();
-                            log(String.format("REV-INSERT: Update order from %d to %d for %s",
-                                (orderId == null ? -1 : orderId),
+                Tuple tuple = new Tuple(rev, addressLinesId, line, revType, orderId, rowNumber);
+                tuples.add(tuple);
+            }
+
+            if (!processTuples(tuples, updateOrder))
+            {
+                log("Re-Order addresslines_aud ID = " + tuples.get(0).getAddressLinesId() + " and try again.");
+                if (processTuples(reOrderTuples(tuples), updateOrder))
+                    log("!!Success!!");
+                else
+                    log("!!Still failed!! Need manual fix for addresslines_aud ID = "
+                            + tuples.get(0).getAddressLinesId());
+            }
+
+        }
+    }
+
+    private static boolean processTuples(ArrayList<Tuple> tupleList, PreparedStatement updateOrder)
+    {
+        final HashMap<Tuple, Integer> currentHistoryMap = new LinkedHashMap<>();
+        final ArrayList<Tuple> currentTuples = new ArrayList<>();
+        int order;
+
+        for (Tuple tuple : tupleList) {
+            int revType = tuple.getRevType();
+            switch (revType) {
+                case 0:
+                    currentTuples.add(tuple);
+                    order = currentTuples.size() - 1; // 0 based
+                    // before we add a new line, we need to check if the new line order has already being taken
+                    // by any previous existing line
+                    if (order > 0 &&
+                            currentTuples.stream().anyMatch(tuple1 -> (currentHistoryMap.get(tuple1) !=
+                                    null && currentHistoryMap.get(tuple1) == currentTuples.size() - 1))) {
+                        log("REV-DELETE: !!We guessed the wrong order!! " +
+                                "We need to manually fix addresslines ID" + tuple);
+                        return false;
```

Comment:
Probably. Best case the order_id at least reflects the original insertion order. Worst case the order_id is messed up regardless of whether you rolled back.